README.md
# sha1

[![Build status](https://travis-ci.org/mpeterv/sha1.png?branch=master)](https://travis-ci.org/mpeterv/sha1)
[![Test coverage](https://codecov.io/gh/mpeterv/sha1/branch/master/graph/badge.svg)](https://codecov.io/gh/mpeterv/sha1)
[![License](https://img.shields.io/badge/License-MIT-brightgreen.svg)](LICENSE)

This module implements SHA-1 and HMAC-SHA-1 in pure Lua. For better performance it uses different bitwise operation implementations depending on environment:

* On Lua 5.1:
  - Uses `bit` module provided by [luabitop](https://luarocks.org/modules/luarocks/luabitop) rock if it is available.
  - Otherwise, uses `bit32` module provided by [bit32](https://luarocks.org/modules/siffiejoe/bit32) rock if it is available.
  - Otherwise, uses an implementation written in Lua.
* On Lua 5.2: uses built-in `bit32` module.
* On Lua 5.3: uses built-in bitwise operators.
* On LuaJIT 2.x: uses built-in `bit` module.

## Installation

For installation using [LuaRocks](https://luarocks.org/) run `luarocks install sha1`.

For manual installation copy `src/sha1` into a directory within `package.path`.

## Usage

```lua
local sha1 = require "sha1"

-- Prints module version in MAJOR.MINOR.PATCH format.
print(sha1.version)

-- Returns a hex string of length 40. sha1(message) also works.
local hash_as_hex = sha1.sha1(message)

-- Returns raw bytes as a string of length 20.
local hash_as_data = sha1.binary(message)

-- Returns a hex string of length 40.
local hmac_as_hex = sha1.hmac(key, message)

-- Returns raw bytes as a string of length 20.
local hmac_as_data = sha1.hmac_binary(key, message)
```

## Benchmarking

Run `lua bench.lua` to benchmark `sha` (requires [argparse](https://github.com/mpeterv/argparse) and [luasocket](http://w3.impa.br/~diego/software/luasocket/)). See `lua bench.lua -h` for more options.

Example results, running on a machine with Intel Core i7-7700HQ and DDR4 SDRAM:

```
Lua 5.1.5
Start up: 0.016858 seconds
SHA-1(1000 characters) 1000 times: 2.121885 seconds

Lua 5.1.5 with luabitop installed:
Start up: 0.001485 seconds
SHA-1(1000 characters) 1000 times: 0.459673 seconds

Lua 5.1.5 with bit32 installed:
Start up: 0.001301 seconds
SHA-1(1000 characters) 1000 times: 0.497699 seconds

Lua 5.2.4
Start up: 0.000355 seconds
SHA-1(1000 characters) 1000 times: 0.529025 seconds

Lua 5.3.5
Start up: 0.000293 seconds
SHA-1(1000 characters) 1000 times: 0.440408 seconds

LuaJIT 2.0.5
Start up: 0.000337 seconds
SHA-1(1000 characters) 1000 times: 0.029341 seconds

LuaJIT 2.1.0-beta3
Start up: 0.000279 seconds
SHA-1(1000 characters) 1000 times: 0.028905 seconds
```

## Testing

To run the test suite ensure that [busted](http://olivinelabs.com/busted/) testing framework is installed and run `busted`.

## Credits

`sha1` module is built upon an implementation by Eike Decker, based on original work by Jeffrey Friedl and cleaned up by [Enrique García Cota](https://github.com/kikito).

## License

This version of the module, as well as all the previous ones on which it is based, are implemented under the MIT license (see LICENSE file for details).


CHANGELOG.md
# Changelog

## 0.6.0 (unreleased)

### Added

* New `sha1.version` field contains `sha1` version.

### Changed

* `sha1` now uses bitwise operation modules or Lua 5.3 bitwise operators when
  available. This improves performance considerably. Pure Lua fallback
  used on Lua 5.1 is now much faster, too.

## 0.5.0 (2014-01-22)

Initial LuaRocks release.


bench.lua
local argparse = require "argparse"
local socket = require "socket"

local function bench_fn(label, fn)
   local start_time = socket.gettime()
   fn()
   local finish_time = socket.gettime()
   print(("%s: %f seconds"):format(label, finish_time - start_time))
end

local input_strings = {}

local function gen_input_strings(string_length, num_strings)
   for _ = 1, num_strings do
      local chars = {}

      for _ = 1, string_length do
         table.insert(chars, string.char(math.random(0, 255)))
      end

      table.insert(input_strings, table.concat(chars))
   end
end

local cli = argparse("bench.lua", "sha1 benchmarking script.")
cli:option("-c --count", "Number of strings to run sha1.sha1 on.", "1000", tonumber)
cli:option("-l --length", "Length of each string.", "1000", tonumber)
cli:option("-s --startups", "Number of times to load sha1.", "100", tonumber)
cli:option("-m --module", "Module to use on Lua 5.1 (bit or bit32).", nil, {bit = "bit", bit32 = "bit32"})

local args = cli:parse()

if _VERSION:find("5%.[12]") and not _G.jit then
   if args.module then
      if not pcall(require, args.module) then
         cli:error(("%s module is not available"):format(args.module))
      end

      if not _VERSION:find("5%.2") and args.module == "bit32" then
         package.loaded[args.module] = nil
      end
   end

   if args.module ~= "bit" then
      package.preload.bit = error
   end

   if args.module ~= "bit32" then
      package.preload.bit32 = error
   end
end

local sha1

bench_fn(("Start up %d times"):format(args.startups), function()
   for _ = 1, args.startups do
      sha1 = require "sha1"

      for module_name in pairs(package.loaded) do
         if module_name:find("^sha1") then
            package.loaded[module_name] = nil
         end
      end
   end
end)

gen_input_strings(args.length, args.count)

bench_fn(("SHA-1(%d characters) %d times"):format(args.length, args.count), function()
   for _, input_string in ipairs(input_strings) do
      sha1.sha1(input_string)
   end
end)


src/sha1/pure_lua_ops.lua
local common = require "sha1.common"

local ops = {}

local bytes_to_uint32 = common.bytes_to_uint32
local uint32_to_bytes = common.uint32_to_bytes

function ops.uint32_lrot(a, bits)
   local power = 2 ^ bits
   local inv_power = 0x100000000 / power
   local lower_bits = a % inv_power
   return (lower_bits * power) + ((a - lower_bits) / inv_power)
end

-- Build caches for bitwise `and` and `xor` over bytes to speed up uint32 operations.

-- Returns a cache containing all values of a 8bit bitwise operator, given an operator over single bits.
-- Value of `op(a, b)` is stored in `cache[a * 256 + b]`.
local function make_op_cache(bit_op)
   local bit_op_0_0 = bit_op(0, 0)
   local bit_op_0_1 = bit_op(0, 1)
   local bit_op_1_0 = bit_op(1, 0)
   local bit_op_1_1 = bit_op(1, 1)

   local cache = {[0] = bit_op_0_0}

   -- The implementation used is an optimized version of the following reference one,
   -- with intermediate calculations reused and moved to the outermost loop possible,
   -- and with multiplications by loop counter replaced with incremental additions.

   -- for a = 0, 127 do
   --    for b = 0, 127 do
   --       -- For all pairs of 7bit numbers (a, b) (so, both a and b have their highest bit clear):
   --       -- Let v be (a op b) << 1; then the following equations hold:
   --       -- * (a << 1) op (b << 1) = v + (0 op 0);
   --       -- * (a << 1) op ((b << 1) + 1) = v + (0 op 1);
   --       -- * ((a << 1) + 1) op (b << 1) = v + (1 op 0);
   --       -- * ((a << 1) + 1) op ((b << 1) + 1) = v + (1 op 1);
   --       local v = cache[y * 256 + x] * 2
   --       cache[a * 2 * 256 + b * 2] = v + bit_op_0_0
   --       cache[a * 2 * 256 + b * 2 + 1] = v + bit_op_0_1
   --       cache[(a * 2 + 1) * 256 + b * 2] = v + bit_op_1_0
   --       cache[(a * 2 + 1) * 256 + b * 2 + 1] = v + bit_op_1_1
   --       -- a op b is cached on the step with (a', b') = (a >> 1, b >> 1),
   --       -- so that it's always defined by the time it's needed (a = 0, b = 0 is the base case).
   --    end
   -- end

   local a_by_2_by_256_plus_b_by_2 = 0

   for a_by_256 = 0, 127 * 256, 256 do
      for a_by_256_plus_b = a_by_256, a_by_256 + 127 do
         local v = cache[a_by_256_plus_b] * 2
         cache[a_by_2_by_256_plus_b_by_2] = v + bit_op_0_0
         cache[a_by_2_by_256_plus_b_by_2 + 1] = v + bit_op_0_1
         cache[a_by_2_by_256_plus_b_by_2 + 256] = v + bit_op_1_0
         cache[a_by_2_by_256_plus_b_by_2 + 257] = v + bit_op_1_1
         a_by_2_by_256_plus_b_by_2 = a_by_2_by_256_plus_b_by_2 + 2
      end

      a_by_2_by_256_plus_b_by_2 = a_by_2_by_256_plus_b_by_2 + 256
   end

   return cache
end

local byte_and_cache = make_op_cache(function(a, b) return a * b end)
local byte_xor_cache = make_op_cache(function(a, b) return a == b and 0 or 1 end)

function ops.byte_xor(a, b)
   return byte_xor_cache[a * 256 + b]
end

function ops.uint32_xor_3(a, b, c)
   local a1, a2, a3, a4 = uint32_to_bytes(a)
   local b1, b2, b3, b4 = uint32_to_bytes(b)
   local c1, c2, c3, c4 = uint32_to_bytes(c)

   return bytes_to_uint32(
      byte_xor_cache[a1 * 256 + byte_xor_cache[b1 * 256 + c1]],
      byte_xor_cache[a2 * 256 + byte_xor_cache[b2 * 256 + c2]],
      byte_xor_cache[a3 * 256 + byte_xor_cache[b3 * 256 + c3]],
      byte_xor_cache[a4 * 256 + byte_xor_cache[b4 * 256 + c4]]
   )
end

function ops.uint32_xor_4(a, b, c, d)
   local a1, a2, a3, a4 = uint32_to_bytes(a)
   local b1, b2, b3, b4 = uint32_to_bytes(b)
   local c1, c2, c3, c4 = uint32_to_bytes(c)
   local d1, d2, d3, d4 = uint32_to_bytes(d)

   return bytes_to_uint32(
      byte_xor_cache[a1 * 256 + byte_xor_cache[b1 * 256 + byte_xor_cache[c1 * 256 + d1]]],
      byte_xor_cache[a2 * 256 + byte_xor_cache[b2 * 256 + byte_xor_cache[c2 * 256 + d2]]],
      byte_xor_cache[a3 * 256 + byte_xor_cache[b3 * 256 + byte_xor_cache[c3 * 256 + d3]]],
      byte_xor_cache[a4 * 256 + byte_xor_cache[b4 * 256 + byte_xor_cache[c4 * 256 + d4]]]
   )
end

function ops.uint32_ternary(a, b, c)
   local a1, a2, a3, a4 = uint32_to_bytes(a)
   local b1, b2, b3, b4 = uint32_to_bytes(b)
   local c1, c2, c3, c4 = uint32_to_bytes(c)

   -- (a & b) + (~a & c) has less bitwise operations than (a & b) | (~a & c).
   return bytes_to_uint32(
      byte_and_cache[b1 * 256 + a1] + byte_and_cache[c1 * 256 + 255 - a1],
      byte_and_cache[b2 * 256 + a2] + byte_and_cache[c2 * 256 + 255 - a2],
      byte_and_cache[b3 * 256 + a3] + byte_and_cache[c3 * 256 + 255 - a3],
      byte_and_cache[b4 * 256 + a4] + byte_and_cache[c4 * 256 + 255 - a4]
   )
end

function ops.uint32_majority(a, b, c)
   local a1, a2, a3, a4 = uint32_to_bytes(a)
   local b1, b2, b3, b4 = uint32_to_bytes(b)
   local c1, c2, c3, c4 = uint32_to_bytes(c)

   -- (a & b) + (c & (a ~ b)) has less bitwise operations than (a & b) | (a & c) | (b & c).
   return bytes_to_uint32(
      byte_and_cache[a1 * 256 + b1] + byte_and_cache[c1 * 256 + byte_xor_cache[a1 * 256 + b1]],
      byte_and_cache[a2 * 256 + b2] + byte_and_cache[c2 * 256 + byte_xor_cache[a2 * 256 + b2]],
      byte_and_cache[a3 * 256 + b3] + byte_and_cache[c3 * 256 + byte_xor_cache[a3 * 256 + b3]],
      byte_and_cache[a4 * 256 + b4] + byte_and_cache[c4 * 256 + byte_xor_cache[a4 * 256 + b4]]
   )
end

return ops


src/sha1/init.lua
local common = require "sha1.common"

local sha1 = {
   -- Meta fields retained for compatibility.
   _VERSION     = "sha.lua 0.6.0",
   _URL         = "https://github.com/mpeterv/sha1",
   _DESCRIPTION = [[
SHA-1 secure hash and HMAC-SHA1 signature computation in Lua,
using bit and bit32 modules and Lua 5.3 operators when available
and falling back to a pure Lua implementation on Lua 5.1.
Based on code orignally by Jeffrey Friedl and modified by
Eike Decker and Enrique García Cota.]],
   _LICENSE = [[
MIT LICENSE

Copyright (c) 2013 Enrique García Cota, Eike Decker, Jeffrey Friedl
Copyright (c) 2018 Peter Melnichenko

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.]]
}

sha1.version = "0.6.0"

local function choose_ops()
   if _VERSION:find("5%.3") then
      return "lua53_ops"
   elseif pcall(require, "bit") then
      return "bit_ops"
   elseif pcall(require, "bit32") then
      return "bit32_ops"
   else
      return "pure_lua_ops"
   end
end

local ops = require("sha1." .. choose_ops())
local uint32_lrot = ops.uint32_lrot
local byte_xor = ops.byte_xor
local uint32_xor_3 = ops.uint32_xor_3
local uint32_xor_4 = ops.uint32_xor_4
local uint32_ternary = ops.uint32_ternary
local uint32_majority = ops.uint32_majority

local bytes_to_uint32 = common.bytes_to_uint32
local uint32_to_bytes = common.uint32_to_bytes

local sbyte = string.byte
local schar = string.char
local sformat = string.format
local srep = string.rep

local function hex_to_binary(hex)
   return (hex:gsub("..", function(hexval)
      return schar(tonumber(hexval, 16))
   end))
end

-- Calculates SHA1 for a string, returns it encoded as 40 hexadecimal digits.
function sha1.sha1(str)
   -- Input preprocessing.
   -- First, append a `1` bit and seven `0` bits.
   local first_append = schar(0x80)

   -- Next, append some zero bytes to make the length of the final message a multiple of 64.
   -- Eight more bytes will be added next.
   local non_zero_message_bytes = #str + 1 + 8
   local second_append = srep(schar(0), -non_zero_message_bytes % 64)

   -- Finally, append the length of the original message in bits as a 64-bit number.
   -- Assume that it fits into the lower 32 bits.
   local third_append = schar(0, 0, 0, 0, uint32_to_bytes(#str * 8))

   str = str .. first_append .. second_append .. third_append
   assert(#str % 64 == 0)

   -- Initialize hash value.
   local h0 = 0x67452301
   local h1 = 0xEFCDAB89
   local h2 = 0x98BADCFE
   local h3 = 0x10325476
   local h4 = 0xC3D2E1F0

   local w = {}

   -- Process the input in successive 64-byte chunks.
   for chunk_start = 1, #str, 64 do
      -- Load the chunk into W[0..15] as uint32 numbers.
      local uint32_start = chunk_start

      for i = 0, 15 do
         w[i] = bytes_to_uint32(sbyte(str, uint32_start, uint32_start + 3))
         uint32_start = uint32_start + 4
      end

      -- Extend the input vector.
      for i = 16, 79 do
         w[i] = uint32_lrot(uint32_xor_4(w[i - 3], w[i - 8], w[i - 14], w[i - 16]), 1)
      end

      -- Initialize hash value for this chunk.
      local a = h0
      local b = h1
      local c = h2
      local d = h3
      local e = h4

      -- Main loop.
      for i = 0, 79 do
         local f
         local k

         if i <= 19 then
            f = uint32_ternary(b, c, d)
            k = 0x5A827999
         elseif i <= 39 then
            f = uint32_xor_3(b, c, d)
            k = 0x6ED9EBA1
         elseif i <= 59 then
            f = uint32_majority(b, c, d)
            k = 0x8F1BBCDC
         else
            f = uint32_xor_3(b, c, d)
            k = 0xCA62C1D6
         end

         local temp = (uint32_lrot(a, 5) + f + e + k + w[i]) % 0x100000000
         e = d
         d = c
         c = uint32_lrot(b, 30)
         b = a
         a = temp
      end

      -- Add this chunk's hash to result so far.
      h0 = (h0 + a) % 0x100000000
      h1 = (h1 + b) % 0x100000000
      h2 = (h2 + c) % 0x100000000
      h3 = (h3 + d) % 0x100000000
      h4 = (h4 + e) % 0x100000000
   end

   return sformat("%08x%08x%08x%08x%08x", h0, h1, h2, h3, h4)
end

function sha1.binary(str)
   return hex_to_binary(sha1.sha1(str))
end

-- Precalculate replacement tables.
local xor_with_0x5c = {}
local xor_with_0x36 = {}

for i = 0, 0xff do
   xor_with_0x5c[schar(i)] = schar(byte_xor(0x5c, i))
   xor_with_0x36[schar(i)] = schar(byte_xor(0x36, i))
end

-- 512 bits.
local BLOCK_SIZE = 64

function sha1.hmac(key, text)
   if #key > BLOCK_SIZE then
      key = sha1.binary(key)
   end

   local key_xord_with_0x36 = key:gsub('.', xor_with_0x36) .. srep(schar(0x36), BLOCK_SIZE - #key)
   local key_xord_with_0x5c = key:gsub('.', xor_with_0x5c) .. srep(schar(0x5c), BLOCK_SIZE - #key)

   return sha1.sha1(key_xord_with_0x5c .. sha1.binary(key_xord_with_0x36 .. text))
end

function sha1.hmac_binary(key, text)
   return hex_to_binary(sha1.hmac(key, text))
end

setmetatable(sha1, {__call = function(_, str) return sha1.sha1(str) end})

return sha1
